### ISSUE TITLE
Rewrite middleware does not write multiple EDNS0 options

### ISSUE THREAD

**Issue by huynhmb359** [coredns/coredns#936]
In the function Rewrite.ServeDNS(),  the loop that iterates through the list of rules returns when the rule.Rewrite() returns RewriteDone.    The rule.Rewrite() always returns RewriteDone when it successfully completes the rewrite.   This means that the next rule in the rule list will not be executed.  

I would expect it inserted all the EDNS0 in the rule list. 
I did a tcpdump and verify that it always set/replace the first rule only.

**Comment by johnbelamaric** [comment]
Can you show me your Corefile? At least the rewrite part.

**Comment by huynhmb359** [comment]
cat Corefile 
.:53 {
  proxy . 10.82.3.15:53 

  rewrite edns0 local set 0xfff1 0xaabbccdd
  rewrite edns0 local set 0xfff2 0x111111
  rewrite edns0 local set 0xfff3 $client_ip
  rewrite edns0 local set 0xfff4 $protocol
 
  log stdout
  errors stdout
}

**Comment by johnbelamaric** [comment]
Ok. That should work, so it's certainly a bug. I thought we had this working already though and had used it.

**Comment by johnbelamaric** [comment]
I believe it used to insert *multiple* rewrite middleware when configured this way.

**Comment by huynhmb359** [comment]
In the ServeDNS() function, 
```
 for _, rule := range rw.Rules {
		switch result := rule.Rewrite(w, r); result {
		case RewriteDone:
			if rw.noRevert {
				return middleware.NextOrFailure(rw.Name(), rw.Next, ctx, w, r)
			}
			return middleware.NextOrFailure(rw.Name(), rw.Next, ctx, wr, r)
		case RewriteIgnored:
			break
		case RewriteStatus:
			// only valid for complex rules.
			// if cRule, ok := rule.(*ComplexRule); ok && cRule.Status != 0 {
			// return cRule.Status, nil
			// }
		}
	}
	return middleware.NextOrFailure(rw.Name(), rw.Next, ctx, w, r)
```
As  the first rule completes Rewrite() function and returns RewriteDone,  the function ServeDNS() returns.    There is no execution of the next rule's Rewrite().  

To fix this we need to know the behaviour when some rules in the list return RewriteDone and some return  RewriteIgnored.

**Comment by miekg** [comment]
This was lifted (at the time) from Caddy and made to work in CoreDNS. What do complex rules actually do? I.e. do *they* allow for chaining or more fancy rewriting (if adapted to coredns)?

**Comment by johnbelamaric** [comment]
When you specify multiple rewrite lines, it is collecting those into a single rewrite middleware instance with different rules? Is that the expected behavior?

**Comment by huynhmb359** [comment]
Yes in the current code,  it collects multiple write lines into a single rewrite middleware.
"Expected behavior",  we need input from @miekg .

**Comment by johnbelamaric** [comment]
Ok. I didn't think it used to do that. But I could be wrong.

**Comment by johnbelamaric** [comment]
We may need a way for a rule to indicate whether it stops the chain or not.

**Comment by miekg** [comment]
I can't following everything that closely:

a) problem is chaining queries?
b) what do you propose and what pros and cons do you see?
c) what syntax should this have?
c) time for a new middleware?

**Comment by johnbelamaric** [comment]
The issue is just having more than one rule apply. So, you can rewrite the same query with two rules. Currently it doesn't do that. I thought that in the past it did, because I thought that for each `rewrite` directive in the Corefile, it would install a separate `rewrite` middleware in the chain. But that's not the case. I guess I just wasn't thinking it through...

So, we need to have a way for a given rule to indicate whether or not the processing should continue. Maybe an optional argument just before the FIELD. Instead of:
```
rewrite FIELD FROM TO
```

we would have

```
rewrite [continue|stop] FIELD FROM TO
```

with the default being `stop` which is the current behavior.

**Comment by miekg** [comment]
Fixed via #1096
